### PR TITLE
Manual grading: separate points/score for manual and automated grading

### LINKED ISSUE #3191: Continue to allow resubmissions after scoring 100
It seems that when students get 100 on a coding problem, they can't submit anymore. This has caused various issues for students psychologically or for staff logistically (students can't resubmit their _preferred_ version, staff can't refer to the version the student _meant_ to submit, etc.). If a question has manual review involved, this can be a problem.

Can we add an option to allow continued saving and submitting even when 100 is achieved?

(I thought there was already a feature request open for this, but I can't find it now.)

### LINKED ISSUE #5110: Manually Graded Questions are Marked as 'External' grading even when done in PrairieLearn
The sproc `instance_questions_update_score`, marks every question that its called on as `grading_method='External'`, even though it is often used to update scores from within PrairieLearn, both for the manual score overrides that have been around for a while, and for the new dedicated manual grading user interface.

https://github.com/PrairieLearn/PrairieLearn/blob/0473cb9a848c5b93f3803f371da83fc12e7eac81/sprocs/instance_questions_update_score.sql#L126-L151

### LINKED ISSUE #5915: docs mention a "reset" assessment button?
https://prairielearn.readthedocs.io/en/latest/assessment/#assessment-and-question-instances-and-resetting-assessments mentions 

> However, if you want to force the regeneration of assessment instances then you can do so with the "reset" button on instructor view of the assessment. While writing an assessment you might need to do this many times

Did the code get refactored to replace the "reset" button with something else and the docs didn't get updated? Or is that the "... all instances" dropdown on the assessment Students page?

### LINKED ISSUE #6126: Allow tagging as "requires grading" in manual grading page
I would like to be able to tag a submission as "Requires grading" from this menu:

image

### PR DESCRIPTION
**Pull request by jonatanschroeder**
This PR enables the separation of points between manual grading and automated grading. The main purpose of this separation is to enable instructors to have questions with automated results, but with a component that is graded manually, such as:

* calculations that include a section to show work;
* code quality (comments, style, etc.);
* answers that require an explanation.

Tasks:

* [x] database;
* [x] schema;
* [x] sync;
* [x] sync tests;
* [x] sample question in exampleCourse;
* [x] review sprocs/libs that update scores;
* [x] review sprocs/libs that retrieve scores;
* [x] UI for instructor:
    * [x] list of questions;
    * [x] manual grading overview;
    * [x] assessment question;
    * [x] instance question;
    * [x] downloads;
    * [x] file upload;
    * [x] question stats;
    * [x] assessment instance log;
* [x] UI for student:
    * [x] assessment overview;
    * [x] instance question;
* [x] UI tests;
* [x] API entrypoints;
* [x] remove all old references to _in_grading;
* [x] Resolve #5110 
* [x] Resolve #6126
* [x] Resolve #5915 
* [x] Resolve #3191
* [x] Docs
* [x] News item

### CODE REVIEW

File: news_items/040_manual_grading_ui/info.json
Comment (jonatanschroeder) [inline news_items/040_manual_grading_ui/info.json:4]: @mwest1066 I listed you as author based on previous discussion of having news items be authored by someone else, let me know if you'd like me to replace with someone else instead.

File: docs/question.md
Comment (jonatanschroeder) [inline docs/question.md:280]: I removed this comment since it is no longer true, but if you believe it's better to replace it with a comment on setting manual points I'm happy to do it.

File: sprocs/assessment_instances_grade.sql
Comment (nwalters512) [inline sprocs/assessment_instances_grade.sql:20]: It looks as though `points_in_grading` and `score_perc_in_grading` are no longer used anywhere, is that correct? If so, can you open a ticket to remove all associated columns once this PR has been deployed?

File: sprocs/instance_questions_update_in_grading.sql
Comment (nwalters512) [inline sprocs/instance_questions_update_in_grading.sql:2]: Can you explain why this sproc is no longer needed?

**Review by nwalters512 — COMMENTED** [review]
I took a first pass through. I'll admit my eyes kind of glazed over after a while 😅 if there's anything specific you'd like feedback on, please call it out!

File: docs/c-grader/index.md
Comment (jonatanschroeder) [reply docs/c-grader/index.md:383]: This has been updated here:

https://github.com/PrairieLearn/PrairieLearn/blob/b47d29913a313e0587c475912586184957db1549/sprocs/instance_questions_points_exam.sql#L45-L54

File: docs/manualGrading.md
Comment (jonatanschroeder) [reply docs/manualGrading.md:10]: > To check my own understanding: will setting `manualPoints` without also setting `autoPoints` function basically the same as `"gradingMethod": "Manual"`?

Yes, that is correct, though setting the grading method to `Manual` still can be used to:
* set the default if `points` is used instead of `manualPoints`;
* change the default behaviour of the question preview page.

> If it can, I'm wondering if we should break up the docs into the "preferred" manual grading flow (all your great new work) and the "legacy"/"advanced" manual grading workflow. If someone is just skimming this page, I'd want them to hit the modern process and not waste time with `gradingMethod`/`singleVariant`.

That can be done, though all example course questions continue to have references to points and the manual grading method. Should that be kept? Also, there's still use for the suggestion to use `singleVariant`.

File: docs/manualGrading.md
Comment (jonatanschroeder) [reply docs/manualGrading.md]: I reworded it, is it clearer now?

File: lib/question.js
Comment (jonatanschroeder) [reply lib/question.js:518]: Yes, that is correct, and also because a question with other grading methods may be configured with manualPoints only. This logic has now been moved to `variants_select_submission_for_grading` instead.

https://github.com/PrairieLearn/PrairieLearn/blob/74db1d9681db56bb83d48ce888c375784251abe3/sprocs/variants_select_submission_for_grading.sql#L26-L35

File: pages/instructorAssessmentDownloads/instructorAssessmentDownloads.js
Comment (jonatanschroeder) [reply pages/instructorAssessmentDownloads/instructorAssessmentDownloads.js:225]: What I've done here is I removed some columns that are not useful in the download/upload grading process (name and role). These columns weren't part of the query anyway, so their value was always blank. Given the additional columns added by this PR I took the opportunity to remove them here as well. Given the conflict this would generate with this PR (this change would also imply changes in the docs) I'd prefer to keep this together.

File: pages/instructorAssessmentDownloads/instructorAssessmentDownloads.sql
Comment (jonatanschroeder) [reply pages/instructorAssessmentDownloads/instructorAssessmentDownloads.sql]: > What does "for the extra" mean?

If `max_auto_points` is 5, but the question has 7 points, 2 points (the extra) are set to manual. I removed it for clarity.

File: migrations/20220816180608_manual_points__add.sql
Comment (jonatanschroeder) [reply migrations/20220816180608_manual_points__add.sql:13]: As mentioned in Slack, will be reused for rubrics in #6271.

File: tests/manualGrading.test.js
Comment (jonatanschroeder) [reply tests/manualGrading.test.js:567]: There are multiple elements like this in the same block, for manual points and auto points. This one is explicitly selecting the total points.

File: sprocs/assessment_instances_grade.sql
Comment (jonatanschroeder) [reply sprocs/assessment_instances_grade.sql:20]: Done in #5911

File: sprocs/instance_questions_update_in_grading.sql
Comment (jonatanschroeder) [reply sprocs/instance_questions_update_in_grading.sql:2]: It was never used to begin with. Also, the `_in_grading` fields were updated, but never used in any API, UI or meaningful result. There's no clear documentation as to what these fields were supposed to represent (I believe it's the number of points that can still be achievable, but there are other better ways to compute that).

File: docs/manualGrading.md
Comment (jonatanschroeder) [reply docs/manualGrading.md]: Updated docs. Grading method has been removed from this page, left only in other pages. I also changed the assessment configuration docs.

File: docs/assessment.md
Comment (jonatanschroeder) [reply docs/assessment.md:115]: I added documentation for the current behaviour, but based on #3618 this may be changed to opt-in instead of opt-out in the future.

File: docs/c-grader/index.md
Comment (nwalters512) [reply docs/c-grader/index.md:383]: Gotcha, thanks for the link.

File: docs/manualGrading.md
Comment (nwalters512) [reply docs/manualGrading.md]: It is!

File: pages/instructorAssessmentDownloads/instructorAssessmentDownloads.js
Comment (nwalters512) [reply pages/instructorAssessmentDownloads/instructorAssessmentDownloads.js:225]: Makes sense, thanks for providing the extra context.

**Review by nwalters512 — APPROVED** [review]
This looks good to me, though given how many bits of PL this touches, I'd also like to get an approval from @mwest1066 before we merge this!

File: pages/instructorAssessmentInstance/instructorAssessmentInstance.sql
Comment (mwest1066) [inline pages/instructorAssessmentInstance/instructorAssessmentInstance.sql:74]: I'm finding the renaming here (and similar renaming elsewhere) to be a bit confusing. I _think_ I understand why we are doing it. Would it be cleaner to rename `aq.manual_points` to `aq.max_manual_points` to match `aq.max_auto_points`? This would match the naming in `question_score_logs`.

File: pages/instructorAssessmentInstance/instructorAssessmentInstance.sql
Comment (mwest1066) [inline pages/instructorAssessmentInstance/instructorAssessmentInstance.sql:64]: Do we have a plan to clean this up in the DB? I'd be happy to run a migration that does this update and then we can eliminate this code. I'm worried that we have many copies of this code block (and comment) but in the future it will be easy to forget this.

File: migrations/20220816180621_assessment_questions__manual_points__update.sql
Comment (mwest1066) [reply migrations/20220816180621_assessment_questions__manual_points__update.sql:11]: Yep. @jonatanschroeder you can remind us about this?

File: news_items/040_manual_grading_ui/index.html
Comment (mwest1066) [inline news_items/040_manual_grading_ui/index.html:57]: Is this what was meant here?
```suggestion
  interface, allowing course staff to provide feedback on individual
```

File: pages/instructorAssessmentInstance/instructorAssessmentInstance.sql
Comment (jonatanschroeder) [reply pages/instructorAssessmentInstance/instructorAssessmentInstance.sql:74]: That's a great point. I'll rename it. Should I also rename it in the schema for infoAssessment.json (`maxManualPoints` instead of `manualPoints`)? I don't think it's necessary, as I imagine most instructors would not use `maxAutoPoints` often and having `manualPoints` matches the `autoPoints` field better.

File: pages/instructorAssessmentInstance/instructorAssessmentInstance.sql
Comment (jonatanschroeder) [reply pages/instructorAssessmentInstance/instructorAssessmentInstance.sql:64]: I can do that, I was just worried that this would be a pretty long migration script. If you're ok with running it (twice, once on canary migration and once on full deployment for the leftovers) I'm happy to create a new script.

File: pages/instructorAssessmentInstance/instructorAssessmentInstance.sql
Comment (mwest1066) [reply pages/instructorAssessmentInstance/instructorAssessmentInstance.sql:74]: I agree with you that it's probably best to keep the input variable as `manualPoints` in `infoAssessment.json`.

File: pages/instructorAssessmentInstance/instructorAssessmentInstance.sql
Comment (mwest1066) [reply pages/instructorAssessmentInstance/instructorAssessmentInstance.sql:64]: I think in this case it's worth it to run a long migration to avoid the permanent code complexity in many places. I think the code should assume that the migration has been run so we can simplify all these places where we have code like above.

File: pages/instructorAssessmentInstance/instructorAssessmentInstance.sql
Comment (jonatanschroeder) [reply pages/instructorAssessmentInstance/instructorAssessmentInstance.sql]: Resolved in ae83825.

File: pages/instructorAssessmentInstance/instructorAssessmentInstance.sql
Comment (jonatanschroeder) [reply pages/instructorAssessmentInstance/instructorAssessmentInstance.sql]: Done in d4d0321. I used the same migration file instead of a separate one because (1) both need to run again on deployment, and (2) the instance question update depends on the assessment question update, so they need to run in order.

Comment (mwest1066) [reply pages/instructorAssessmentInstance/instructorAssessmentInstance.sql]: Looks good, thanks!

**Review by mwest1066 — APPROVED** [review]

**Pull request merged**
